File: ml-controller/services/technical_signal_math.py
"""Canonical MACD, parity tested against Worker technicalSignalMath.ts."""
from __future__ import annotations

import math
from collections.abc import Sequence

TECHNICAL_SIGNAL_MATH_VERSION = 'macd-ema-first-seed-12-26-9-v1'


def ema_series(values: Sequence[float], period: int) -> list[float]:
    if len(values) == 0 or period <= 0:
        return []
    alpha = 2.0 / (period + 1)
    result = [float(values[0])]
    for value in values[1:]:
        result.append(float(value)*alpha + result[-1]*(1-alpha))
    return result


def macd_histogram_last(closes: Sequence[float]) -> float | None:
    if len(closes) < 35 or not all(math.isfinite(float(value)) for value in closes):
        return None
    fast, slow = ema_series(closes, 12), ema_series(closes, 26)
    line = [a-b for a,b in zip(fast,slow,strict=True)]
    signal = ema_series(line,9)
    return line[-1]-signal[-1]

File: worker/src/lib/technicalSignalMath.ts
/** Canonical MACD contract: first-observation EMA seed; full signal warmup. */
export const TECHNICAL_SIGNAL_MATH_VERSION = 'macd-ema-first-seed-12-26-9-v1'

export function emaSeries(values: number[], period: number): number[] {
  if (!values.length || period <= 0) return []
  const alpha = 2 / (period + 1)
  const out = [values[0]]
  for (let i = 1; i < values.length; i++) out.push(values[i] * alpha + out[i - 1] * (1 - alpha))
  return out
}

export function macdComponents(closes: number[]): { macd: number; signal: number; histogram: number } | null {
  if (closes.length < 35 || closes.some(value => !Number.isFinite(value))) return null
  const fast = emaSeries(closes, 12)
  const slow = emaSeries(closes, 26)
  const line = fast.map((value, index) => value - slow[index])
  const signals = emaSeries(line, 9)
  const macd = line[line.length - 1]
  const signal = signals[signals.length - 1]
  return { macd, signal, histogram: macd - signal }
}

export function macdHistogramLast(closes: number[]): number | null {
  return macdComponents(closes)?.histogram ?? null
}
